### PR TITLE
feat[pages]: Add custom _routes.json support for Pages Functions projects

### PR DESCRIPTION
**Pull request by CarmenPopoviciu**

### CODE REVIEW

File: fixtures/pages-functions-with-routes-app/tests/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+		let response = await waitUntilReady("http://localhost:8776/greeting/hello");
+		let text = await response.text();
+		expect(text).toMatch("Bonjour le monde!");
+
+		response = await waitUntilReady("http://localhost:8776/greeting/goodbye");
+		text = await response.text();
+		expect(text).toMatch("A plus tard alligator 👋");
+
+		/**
+		 * :sad_panda:
```
Comment (GregBrimble) [inline fixtures/pages-functions-with-routes-app/tests/index.test.ts:74]: Can we make a GitHub issue to track this feature parity in `wrangler pages dev`?

File: packages/wrangler/src/pages/publish.tsx [hunk 2]
```diff
@@ -327,47 +338,66 @@ export const Handler = async ({
...
+				if (err instanceof FatalError) {
+					throw err;
+				}
 			}
-		} catch {}
+		} else {
+			// with generated _routes.json file
+			try {
+				_routes = readFileSync(
+					routesOutputPath as PathOrFileDescriptor,
```
Comment (jahands) [inline packages/wrangler/src/pages/publish.tsx:369]: Are we casting this because `routesOutputPath` might be undefined? If so, I think it might be cleaner to do `else if(routesOutputPath !== undefined){}` on line 365 rather than casting

File: packages/wrangler/src/pages/publish.tsx [hunk 3]
```diff
@@ -389,3 +419,29 @@ export const Handler = async ({
 	);
 	await metrics.sendMetricsEvent("create pages deployment");
 };
+
+/**
+ * Custom _routes.json files need to run through optimization
+ * to potentially reduce the overall worker pipeline size
+ */
+function validateAndOptimizeRoutesFile(
```
Comment (jahands) [inline packages/wrangler/src/pages/publish.tsx:427]: Lets change this to `validateRoutesFile()` and remove the optimization - we have a ticket to stop optimizing custom `_routes.json` files

File: packages/wrangler/src/pages/publish.tsx [hunk 4]
```diff
@@ -312,6 +317,12 @@ export const Handler = async ({
 		_redirects = readFileSync(join(directory, "_redirects"), "utf-8");
 	} catch {}
 
+	try {
+		// Developers can specify a custom _routes.json file, for projects with Pages
+		// Functions or projects in Advanced Mode
+		_routes = readFileSync(join(directory, "_routes.json"), "utf-8");
```
Comment (jahands) [inline packages/wrangler/src/pages/publish.tsx:323]: Maybe we could rename this to `_routesCustom` and the generated routes var to `_routesGenerated` or something? It's kind of hard to follow all the optionality with the same variable name everywhere

**Review by jahands — CHANGES_REQUESTED** [review]
Looking good!!

File: fixtures/pages-functions-with-routes-app/tests/index.test.ts (hunk 1 above)
Comment (CarmenPopoviciu) [reply fixtures/pages-functions-with-routes-app/tests/index.test.ts:74]: +1! that was unexpected for me as well and meant to ask @jahands what we want to do abt it. I think this will create a lot of confusion for users if we leave it as is.

File: packages/wrangler/src/pages/publish.tsx [hunk 5]
```diff
@@ -327,47 +335,65 @@ export const Handler = async ({
...
-			const advancedModeRoutesString = readFileSync(routesPath, "utf-8");
-			const advancedModeRoutes = JSON.parse(advancedModeRoutesString);
-
-			if (!isRoutesJSONSpec(advancedModeRoutes)) {
-				throw new FatalError(
-					"Invalid _routes.json file found at:" + routesPath,
-					1
-				);
-			}
+		if (_routesCustom) {
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/pages/publish.tsx:380]: @jahands what happens if Advanced Mode projects don't provide a custom `_routes.json` file? How are worker "guarded" routes determined?

Comment (jahands) [reply packages/wrangler/src/pages/publish.tsx:380]: If there is no `_routes.json`, don't send one. The backend will default to all routes running through Functions

File: fixtures/pages-functions-with-routes-app/tests/index.test.ts (hunk 1 above)
Comment (jahands) [reply fixtures/pages-functions-with-routes-app/tests/index.test.ts:74]: Created an issue internally, good catch!

File: packages/wrangler/src/pages/publish.tsx [hunk 6]
```diff
@@ -389,3 +415,21 @@ export const Handler = async ({
...
+
+	if (!isRoutesJSONSpec(routes)) {
+		throw new FatalError(
+			`Invalid _routes.json file found at: ${routesPath}. Please make sure the JSON object has the following format:
+      {
+        version: ${ROUTES_SPEC_VERSION};
+        include: string[];
+        exclude: string[];
+      }
+			and that all rules contain at least one entry.
```
Comment (jahands) [inline packages/wrangler/src/pages/publish.tsx:430]: "and that all rules contain at least one entry." isn't quite right - let's change it to "and that at least one include rule is provided."

**Review by jahands — APPROVED** [review]
Great work!
